### PR TITLE
[GSoC] Make Filters persistent in Filter Sheet

### PR DESCRIPTION
**Pull request by viciousAegis**
## Purpose / Description
The current functionality for filters does not allow them to stay persistent, i.e, if I select some filters, apply them, and open the sheet again, those filters get reset. This PR improves on that and allows filters to stay persistent unless explicitly cleared.

Minor refactoring has also been performed.

## Approach
By passing in the `flagSearchItems` list to the `FlagsAdapter` class, It can check whether a filter is selected or not during the view binding, and depending on that, the selected/unselected look is set.

## How Has This Been Tested?

https://user-images.githubusercontent.com/86671025/186949439-7d82618f-f380-4ad0-b280-fe33308c8c33.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 1]
```diff
@@ -215,6 +208,7 @@ class FilterSheetBottomFragment :
 class FlagsAdapter(
     val context: Context?,
     private var dataSet: Array<FilterSheetBottomFragment.Flags>,
+    private val selectedList: MutableList<SearchNode.Flag>,
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:211]: `selectedFlags` or `selectedElements`. The fact that it's a list is already in the type of the value, and is not an information that is very useful to understand the meaning of the content of the list.
Instead, javadoc should indicate that it contains up to date information about selection

Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:211]: Why does it need to be Mutable? Do you plan to use it later?

FYI MutableList inherits from List. So you can take as argument a list and receive a mutable list. This only mean that the current class won't initiate mutation. But the list can be modified elsewhere.

To be honest, I now believe that the onFlagItemClicked could be in the ViewHolder, since it's actually only used there, in which case it would makes sense to have this be Mutable.

Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:211]: I would suggest using a general "collection" instead of a list. While I'm fine with implementing it by a list with only 8 elements, when we go with decks, note type, and tags (if we use them, I don't recall), the set is big enough that having something more efficient would be nice.

And I still hope that will be able to use the same code, abstracting Flags away (i.e. moving them as type parameters) to avoid repeating code. So I'm reviewing this code assuming that instead of Flags we can have types with hundreds or even thousands of values for huge collection

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:211]: Similarly, you could move `isSelected` inside `FlagsAdapter` since it's only used there. This would allow it to take less parameters since it only uses the FlagsAdapter's parameter `flagSearchItems`

**Review by Arthur-Milchior — COMMENTED** [review]
Really great change.
As you change things, I'd appreciate small commits so that I can spend less time reviewing. That I know exactly which part of the code change corresponds to which part of the feature (or that it corresponds to nothing and that it's just cleaning)

As always, I'm really happy to see code that works and do what you want. So my remarks are only here to ensure that we can easily generalize the code to deck, note type, etc... The idea being to eventually be able to replace `Flag` everywhere by a type parameter, and that `FlagsAdapter` is just a specific version of a generic `SearchCategoryAdapter`. Which means thinking about:
* which data structure can hold 8 or a thousand elements
* how to ensure that any methods related to flags are either in this class in order to be automatically generalized

Also, I believe inner classes would improve slightly the readability, by avoiding to repeat some parameters

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt [hunk 2]
```diff
@@ -528,7 +528,6 @@ open class CardBrowser :
 
     fun searchWithFilterQuery(filterQuery: String) {
         mSearchTerms = filterQuery
-
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt:531]: irrelevant line change

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:211]: I'm not sure I understand this. 

A `List` is an interface for an indexed collection, rather than a Linked List, so it shouldn't affect performance unless we specifically do not want indexing

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 3]
```diff
@@ -226,11 +220,18 @@ class FlagsAdapter(
...
             clickListener: OnItemClickListener,
-            position: Int
+            position: Int,
+            selectedList: MutableList<SearchNode.Flag>,
+            context: Context?
         ) {
             item.text = currFlag.getFlagName(itemView.context)
             icon.setImageResource(currFlag.flagIcon)
 
+            if (currFlag.flagNode in selectedList) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:230]: A 1-line comment would be a nice to have here

**Review by david-allison — APPROVED** [review]
I like this. Two optional extensions

* It would be nice if 'Flags' was coloured blue if one or more filters was applied.
* A `@NeedsTest` attribute or two would be useful as documentation

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt]: Got it, apologies!

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 4]
```diff
@@ -79,17 +79,21 @@ class FilterSheetBottomFragment :
         val applyButton = this.findViewById<Button>(R.id.apply_filter_button)
         applyButton.setOnClickListener {
             val filterQuery = createQuery(flagSearchItems)
-
```
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:82]: I'll explain everything here as well as add 1-line comments in the code to make it clearer.
The current behaviour is as follows:

`Apply` button: Create a query with currently selected filters, and close the filter sheet
`Clear` button: Clear the selected filters, search with no filters applied to show all the cards and close the filter sheet
`Cancel` button: Close the filter sheet (cancel the action of selecting filters). The selected filters will remain selected. The same behaviour as clicking out of the filter sheet.

>My understanding is that, now, if you unselected everything, the search would not be applied, and you corrected that? Right? Or does it have other impact?

If everything is unselected, the query sent to `searchCards()` would be `""`, which would result in all the cards being shown, which would be expected behaviour of having not selected any filters. The `Clear` button is basically a shortcut for this.

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 5]
```diff
@@ -139,17 +143,6 @@ class FilterSheetBottomFragment :
         }
     }
 
-    /**
-     * Clear the filters if Bottom Sheet is dismissed
```
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:143]: >Could you clarify what user action corresponding to dismissing? Clicking outside of the sheet? Dragging the sheet down?
Yes, you are correct, both of those constitute dismissing. I will add video of both actions. 
the `Cancel` button calls `dismiss()` so that also does the same thing. And yes, nothing is done in this case

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 6]
```diff
@@ -215,6 +208,7 @@ class FilterSheetBottomFragment :
 class FlagsAdapter(
     val context: Context?,
     private var dataSet: Array<FilterSheetBottomFragment.Flags>,
```
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:210]: I'll update variable names and add javadoc to both of them to make it clearer what they represent

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 1 above)
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:211]: Yes, Set works. I will change it accordingly.

**Comment by viciousAegis** [comment]
> * A `@NeedsTest` attribute or two would be useful as documentation

Where would you like this? I have added it to the following block for now:
```kotlin
if (currFlag.flagNode in selectedItems) {
    itemView.setBackgroundColor(getColorFromAttr(context, R.attr.filterItemBackgroundSelected))
    itemTextView.setTextColor(getColorFromAttr(context, R.attr.filterItemTextColorSelected))
}
```

**Comment by viciousAegis** [comment]
@Arthur-Milchior @david-allison 
All queries have been addressed. This can be squashed into one or two commits as seen fit. 

I will start to move towards abstraction when I start adding the next type of filters, and will address those comments in the commits there.

I just have one query about the behaviour of the following block of code:
https://github.com/ankidroid/Anki-Android/blob/78ecb433cbb31978421a356f84065c450f4696ff/AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.kt#L529-L534

which method causes the keyboard to appear when I call `searchWithFilterQuery()`? Ideally, I wouldn't want the keyboard to come up if I am manually selecting filters.

**Comment by viciousAegis** [comment]
Updated functionality:

https://user-images.githubusercontent.com/86671025/186949541-d4bb1663-63c8-446a-9812-fd0177ab5970.mp4

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt]: It's not clear to who was the question. Personally, I'm fine with set. Simply, I didn't want to forbid the use of list in the case of Flags, and since the specific kind of collection did not seem important, I decided to ask for Collection. 
I actually suspect that List are betters for short sequences, such as flags, as iterating over each element may be faster than hashing, but no big deal one way or another

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 7]
```diff
@@ -231,6 +245,22 @@ class FilterSheetBottomFragment :
...
                     flagSearchItems.remove(item.flagNode)
                 }
+
+                /*
+                Change color of filter header to indicate a filter has been selected
+                 */
+
+                val filterHeader = flagsButton.findViewById<TextView>(R.id.filterByFlagsText)
+                val filterIcon = flagsButton.findViewById<ImageView>(R.id.filter_by_flags_icon)
+                if (flagSearchItems.isNotEmpty()) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:251]: Actually, this only need to be applied if it was not empty and now it is empty. Since changing UI is expensive, and I can imagine a user clicking on many filters one after the other, I'd appreciate if you also add a boolean `wasEmpty` that checked whether it was empty before you add the element; then you'd do `if (wasEmpty) {} else if (flagSearchItems.isEmpty()){}`

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 8]
```diff
@@ -221,6 +216,7 @@ class FilterSheetBottomFragment :
                 icon.setImageResource(currFlag.flagToggleIcon)
 
                 // If flag is currently selected, bind the view with the selected item background and text color
+                @NeedsTest("Test if background color is being correctly set if item is selected")
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:219]: I'd appreciate if you could add it in the CL where the need appeared. (unless it's already merged)

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 9]
```diff
@@ -194,12 +195,6 @@ class FilterSheetBottomFragment :
         return CollectionHelper.getInstance().getCol(activity)
     }
 
-    /**
-     * Add/remove items from list of selected filters
-     * Change background color accordingly
-     * TODO: background color should be retained if selected and swiped down
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:190]: Why did you remove the javadoc? only the todo should be removed I believe. (And you can remove it in the CL where the change occurred)

**Review by Arthur-Milchior — COMMENTED** [review]
A few nit. Thanks for your change

**Comment by Arthur-Milchior** [comment]
To be more clear, as soon as the few changes are accepted, I'm fine to merge.
But, if it's not to hard, I believe it'll be easier to read the history if you make classes internal first (potentially in a first PR that can be reviewed quickly) and then add the new feature. This way, this will avoid to add and remove parameters, and add code to move it immediately after

**Comment by Arthur-Milchior** [comment]
I edited the PR to merge the comment and needTests at the place they should have been. I didn't do the other changes

**Comment by mikehardy** [comment]
@viciousAegis this still appears to have comments to resolve from Arthur? I was scanning through looking for things to merge, I want to merge one of your PRs :sweat_smile:

**Comment by viciousAegis** [comment]
Yes, apologies; got caught up in some assignments the past few days :( I'll get back to Arthur and complete the requested changes tomorrow

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 8 above)
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:223]: There isn't any "need" as such, only that it has been tested on my phone visually, but not run through a unit test to make sure that it in fact is working correctly

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 10]
```diff
@@ -122,13 +125,24 @@ class FilterSheetBottomFragment :
...
 
-        val flagsButton = requireView().findViewById<LinearLayout>(R.id.filterByFlagsText)
-        val flagIcon = requireView().findViewById<ImageView>(R.id.filter_flagListToggle)
+        flagsButton = requireView().findViewById(R.id.filterByFlagsLayout)
+        flagToggleIcon = requireView().findViewById(R.id.filter_flagListToggle)
         val flagsRecyclerViewLayout =
             requireView().findViewById<LinearLayout>(R.id.flagsRecyclerViewLayout)
 
+        if (flagSearchItems.isNotEmpty()) {
+            val filterHeader = flagsButton.findViewById<TextView>(R.id.filterByFlagsText)
```
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:139]: This code is only run the first time the filter sheet is opened, to check if there were any filters that were selected when the sheet was closed last.

**Comment by viciousAegis** [comment]
Still a couple changes needed to address in latest commit

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 11]
```diff
@@ -224,12 +218,19 @@ class FilterSheetBottomFragment :
                 itemTextView.text = currFlag.getFlagName(itemView.context)
                 icon.setImageResource(currFlag.flagIcon)
 
+                // If [currFlag] is currently selected, bind the view with the selected item background and text color
+                @NeedsTest("Test if background color is being correctly set if item is selected")
+                if (isSelected(currFlag)) {
+                    itemView.setBackgroundColor(getColorFromAttr(context, R.attr.filterItemBackgroundSelected))
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:256]: `itemView` is not necessarily a new object. That's the point of recycler view, reusing the view that disappear to put it with new content in a view that appears.
That means that it may have already background color `filterItemBackgroundSelected`. Even if it's not selected. 
So you need to do an `else` and put back the unselected color.

And since it's really the same if/else than in `onFlagItemClicked`, you should probably create an extra function that set the colors and is used by both `onFlagItemClicked` and `bind`

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:139]: I fear a was not clear.
I don't mean you search it multiple time in this function. I meant you search them multiple time in the class. 

Line `232` in `onFlagItemClicked` you have `val filterHeader = flagsButton.findViewById<TextView>(R.id.filterByFlagsText)`
As far as I understand the way this view works, each call will return exactly the same object. So you can put this object as a member of the class instead, this ensure you'll only need to search it only once. 

And same with the view below.

You can do this with the filter icon too

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 11 above)
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:256]: It isn't exactly the same if/else, `onFlagClicked` sets background to selected if it isn't selected previously, while in `bind`, the background is set to selected if it _is_ selected previously, to account for persistence. So they can't be merged into a single function.
I will add an `else` to the `if` statement in `bind`

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:139]: Got it 👍

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 11 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:256]: I beg your pardon, but I believe that you can still create a function.
This function can take a `val selected: Boolean`, that indicates whether the color should indicate selection.
It'll be call with `isSelected(currFlag)` in `bind` and with `!isSelected(item)` in `onFlagItemClicked`

Comment (viciousAegis) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:256]: I see, understood.

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 12]
```diff
@@ -50,15 +49,23 @@ class FilterSheetBottomFragment :
...
+    private var flagSearchItems = mutableSetOf<SearchNode.Flag>()
 
     private lateinit var flagRecyclerView: RecyclerView
     private lateinit var flagListAdapter: FlagsAdapter
 
+    private lateinit var flagsButton: LinearLayout
+    private lateinit var flagToggleIcon: ImageView
+
+    /** Heading of the Flags filter section */
+    private lateinit var filterHeader: TextView
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:61]: Ideally, I'd like the var name to indicate it's related to flag, since you'll probably need similar views for the other categories.
But if we refactor later, that won't be a big deal anyway, so fine by me

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 13]
```diff
@@ -50,15 +49,23 @@ class FilterSheetBottomFragment :
     CollectionGetter {
     private lateinit var behavior: BottomSheetBehavior<View>
 
-    private var flagSearchItems = mutableListOf<SearchNode.Flag>()
+    private var flagSearchItems = mutableSetOf<SearchNode.Flag>()
 
     private lateinit var flagRecyclerView: RecyclerView
     private lateinit var flagListAdapter: FlagsAdapter
 
+    private lateinit var flagsButton: LinearLayout
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:57]: Why "button" while it's a layout. It does not seems to be shaped like a button.

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 14]
```diff
@@ -115,13 +130,24 @@ class FilterSheetBottomFragment :
...
+         *
+         * If a filter is selected, change color of heading
          */
 
-        val flagsButton = requireView().findViewById<LinearLayout>(R.id.filterByFlagsText)
-        val flagIcon = requireView().findViewById<ImageView>(R.id.filter_flagListToggle)
+        flagsButton = requireView().findViewById(R.id.filterByFlagsLayout)
+        filterHeader = flagsButton.findViewById(R.id.filterByFlagsText)
+        filterIcon = flagsButton.findViewById(R.id.filter_by_flags_icon)
+
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:140]: No need for this new line. Seems to indicate the last icon is different from the previous views.

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt (hunk 13 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt:57]: By the way, I don't believe this one was really useful to add as a variable here, since it's used only in `onViewCreated` only, and is executed only once. This is different from the other variables that are used in all click

File: AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt [hunk 15]
```diff
@@ -244,21 +236,45 @@ class FilterSheetBottomFragment :
...
-                }
+                setBackgroundColor(currFlag, true)
 
                 itemView.setOnClickListener {
-                    onFlagItemClicked(currFlag, position)
+                    onFlagItemClicked(currFlag)
+                }
+            }
+
+            fun setBackgroundColor(flag: Flags, setIfSelected: Boolean) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/FilterSheetBottomFragment.kt]: I feel like I was really unclear. In this case, it adds complexity. The goal was to remove some

**Review by Arthur-Milchior — APPROVED** [review]
